homeassistant/components/zha/fan.py

```diff
@@ -74,51 +74,41 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
 class BaseFan(FanEntity):
     """Base representation of a ZHA fan."""
 
-    def __init__(self, *args, **kwargs):
-        """Initialize the fan."""
-        super().__init__(*args, **kwargs)
-        self._state = None
-        self._fan_channel = None
-
-    @property
-    def speed_list(self) -> list:
-        """Get the list of available speeds."""
-        return SPEED_LIST
-
     @property
-    def speed(self) -> str:
-        """Return the current speed."""
-        return self._state
+    def preset_modes(self) -> str:
+        """Return the available preset modes."""
+        return PRESET_MODES
 
     @property
     def supported_features(self) -> int:
         """Flag supported features."""
         return SUPPORT_SET_SPEED
 
-    #
-    # The fan entity model has changed to use percentages and preset_modes
-    # instead of speeds.
-    #
-    # Please review
-    # https://developers.home-assistant.io/docs/core/entity/fan/
-    #
     async def async_turn_on(
         self, speed=None, percentage=None, preset_mode=None, **kwargs
     ) -> None:
         """Turn the entity on."""
-        if speed is None:
-            speed = SPEED_MEDIUM
-
-        await self.async_set_speed(speed)
+        await self.async_set_percentage(percentage)
 
     async def async_turn_off(self, **kwargs) -> None:
         """Turn the entity off."""
-        await self.async_set_speed(SPEED_OFF)
+        await self.async_set_percentage(0)
 
-    async def async_set_speed(self, speed: str) -> None:
-        """Set the speed of the fan."""
-        await self._fan_channel.async_set_speed(SPEED_TO_VALUE[speed])
-        self.async_set_state(0, "fan_mode", speed)
+    async def async_set_percentage(self, percentage: Optional[int]) -> None:
+        """Set the speed percenage of the fan."""
+        if percentage is None:
+            percentage = DEFAULT_ON_PERCENTAGE
+        fan_mode = math.ceil(percentage_to_ranged_value(SPEED_RANGE, percentage))
+        await self._async_set_fan_mode(fan_mode)
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set the speed percenage of the fan."""
+        fan_mode = NAME_TO_PRESET_MODE.get(preset_mode)
```

Comment:
If the preset mode is missing the fan mode will be None. Our typing doesn't match that.

Reply:
Fixed in https://github.com/home-assistant/core/pull/46714